tracking/traffic_info.cpp

```diff
@@ -0,0 +1,106 @@
+#include "tracking/traffic_info.hpp"
+
+#include "platform/http_client.hpp"
+
+#include "coding/bit_streams.hpp"
+#include "coding/reader.hpp"
+#include "coding/write_to_sink.hpp"
+#include "coding/writer.hpp"
+
+#include "base/assert.hpp"
+
+#include "std/algorithm.hpp"
+#include "std/string.hpp"
+
+#include "private.h"
+
+namespace tracking
+{
+namespace
+{
+char const kTestFileName[] = "traffic_data";
+
+bool ReadRemoteFileAsString(string const & url, string & result)
+{
+  platform::HttpClient request(url);
+  if (request.RunHttpRequest() && request.ErrorCode() == 200)
+  {
+    result = request.ServerResponse();
+    return true;
+  }
+  return false;
+}
+}  // namespace
+
+TrafficInfo::TrafficInfo() {}
+
+bool TrafficInfo::ReceiveTrafficData()
+{
+  if (strlen(TRAFFIC_DATA_BASE_URL) == 0)
+    return false;
+
+  string url = string(TRAFFIC_DATA_BASE_URL) + string(kTestFileName);
+  string contents;
+  if (!ReadRemoteFileAsString(url, contents))
+    return false;
+
+  vector<uint8_t> contentBytes(contents.size());
+  for (size_t i = 0; i < contents.size(); ++i)
+    contentBytes[i] = static_cast<uint8_t>(contents[i]);
+
+  DeserializeTrafficData(contentBytes, m_coloring);
+  return true;
+}
+
+SpeedGroup TrafficInfo::GetSpeedGroup(uint32_t id) const
+{
+  auto const it = m_coloring.find(id);
+  if (it == m_coloring.cend())
+    return SpeedGroup::Unknown;
+  return it->second;
+}
+
+// static
+void TrafficInfo::SerializeTrafficData(Coloring const & coloring, vector<uint8_t> & result)
+{
+  uint32_t maxId = 0;
+  for (auto const & p : coloring)
+    maxId = max(maxId, p.first);
+
+  vector<uint8_t> buf;
+  MemWriter<decltype(buf)> memWriter(buf);
+  WriteToSink(memWriter, maxId);
+  {
+    BitWriter<decltype(memWriter)> bitWriter(memWriter);
+    for (uint32_t i = 0; i <= maxId; ++i)
+    {
+      auto const it = coloring.find(i);
+      SpeedGroup g = SpeedGroup::Unknown;
+      if (it != coloring.cend())
+        g = it->second;
+      bitWriter.Write(static_cast<uint8_t>(g), 3);
+    }
+  }
+
+  result.swap(buf);
+}
+
+// static
+void TrafficInfo::DeserializeTrafficData(vector<uint8_t> const & data, Coloring & coloring)
+{
+  MemReader memReader(data.data(), data.size());
+  ReaderSource<decltype(memReader)> src(memReader);
+  auto const maxId = ReadPrimitiveFromSource<uint32_t>(src);
+
+  BitReader<decltype(src)> bitReader(src);
+  coloring.clear();
+  for (uint32_t i = 0; i <= maxId; ++i)
+  {
+    auto const g = static_cast<SpeedGroup>(bitReader.Read(3));
```

Comment:
'3' because all values  are less that SpeedGroup.Count (<= 8), that is only b000...b111), right? may be add a comment?